### DISCUSSION TITLE
Variables in macros for custom commands, shortcuts, piping... etc. How can I find documentation about this?

### DISCUSSION THREAD

**Discussion by borjasanlei** [neomutt/neomutt#4532] [Development]
Probably this question will be obvious or self explanatory for some users. 

As a neomutt user, I would want to create scripts and bind them to shortcats. I'm using macros for this, but I find the documentation on how to use variables for macros unexistant.

I will use a really simple example to showcase the user story I try to fulfill (posted in [stackoverflow](https://stackoverflow.com/questions/77905609/using-variables-in-neomutt-macros-like-current-mail-id))

In neomutt, I want a macro to execute a command, with the current message ID as argument.

As an example:

macro compose <F12> "<shell-escape>mkdir -p ~/tmp/neomutt/<current_id>/<enter>"

where <current_id> would be the desired variable, passing to the macro command the value of the Message ID of the composed email.

How can this be made?

Is it possible to use a variable for the full macro like:

macro compose <F12> "<shell-escape>mkdir -p ~/tmp/neomutt/<current_id>/<enter>\
<copy-file>~/tmp/<current_id>/message.txt<enter>\
<shell-escape>pandoc -o ~/tmp/<current_id>/message.html ~/tmp/<current_id>/message.txt<enter>\

with custom variables?

I've tried different variables like %i. I'd like also to be able to set variables on runtime like NOW=$(time)

The overall goal is to understand neomutt's macros to use variables like 'current email id' or 'sender' or any data available to pipe them into commands or use the built-in set of functions.

Do you know how to perform this? Do you find complete the neomutt documentation in this topic?

**Comment by flatcap** [comment]
Yes it probably can be done, but it'll be tricky.

If you pipe (`<pipe-message>`) the email to a script,
that script could extract the Message ID and perform some other action with it.

**Comment by flatcap** [comment]
tl;dr No, it can't be done.

But it would definitely be useful to be _able_ to do it.

**Note**: Changing this could break existing users' workflows if their commands contained `%x`s.

---

There are five sources of data that we might want:

1. External environment variables
2. Internal environment variables (`setenv`)
3. NeoMutt config variables
4. NeoMutt custom variables (`set my_var = ...`)
5. Expandos for the current email (e.g. `%i` - Message ID)

If I run a NeoMutt command, sources 1, 3 and 4 work:
(This could be `:echo`, `<pipe-message>`, etc)

```
:echo $EXTERNAL $sidebar_width $my_var
```

2. should work and needs fixing.

Getting 5. to work, (probably) _shouldn't_ be too hard.
Pass the current Email (or set of tagged Emails) to the command.
Populate the expandos.
Execute command as before.

We already have lots of code that performs these steps that could be copy'n'pasted.

However, **we'd need a volunteer**.

**Comment by borjasanlei** [comment]
Thanks @flatcap for your time and effort for this issue and for the project.

Unluckly I can not code in C, maybe I can be helpfull with QA or further definitions.

For the index view and other built-in commands, I guess that neomutt already serializes the email into a set of fields (like from, to, id, thread-id...).

Would it be feasable to expose these fields to the macros or even to the CLI? That way neomutt could be completly expandable with custom features as extensions/plugins easily.

Could be done via the number 5. above defined (expandables) or even with a kind of simplified DOM.

I know volunteers are needed, I'd love to contribute with code but I am not able to.

**Comment by borjasanlei** [comment]
Hi @flatcap any further thoughts on this? This is the only feature stopping me from using neomutt since I cannot make custom scripts then.

If that is not on the radar, is there any other way of scripting using data from the emails to pipe or send that into other programs?
